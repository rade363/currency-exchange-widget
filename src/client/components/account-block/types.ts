import { Account } from '../../hooks/useExchange/types';

export type AccountType = 'From' | 'To';

export type OwnProps = {
  type: AccountType;
  currentAccount: Account;
  otherAccount: Account;
  onInputChange: (value: string) => void;
  handleSelectorClick: () => void;
};
