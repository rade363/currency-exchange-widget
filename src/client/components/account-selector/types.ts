import { AccountType } from '../account-block/types';
import { UserAccount } from '../app/types';
import { CurrencyName } from '../../utils/create-currency/types';

export type OwnProps = {
  type: AccountType;
  accounts: UserAccount[];
  closeSelector: () => void;
  selectAccount: (name: CurrencyName) => void;
};
