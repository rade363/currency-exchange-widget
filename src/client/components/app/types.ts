import { CurrencyName } from '../../utils/create-currency/types';

export type User = {
  accounts: UserAccount[];
};

export type UserAccount = {
  balance: number;
  name: CurrencyName;
};
