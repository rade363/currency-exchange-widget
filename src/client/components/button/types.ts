import { ButtonHTMLAttributes } from 'react';

export type OwnProps = {
  active: boolean;
  styleType: 'primary' | 'danger';
} & ButtonHTMLAttributes<HTMLButtonElement>;
