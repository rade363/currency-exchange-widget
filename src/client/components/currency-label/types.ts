import { ButtonHTMLAttributes } from 'react';
import { Currency } from '../../utils/create-currency/types';

export type OwnProps = {
  currency: Currency;
  showArrow?: boolean;
} & ButtonHTMLAttributes<HTMLButtonElement>;
