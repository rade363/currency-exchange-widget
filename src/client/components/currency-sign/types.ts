import { HTMLAttributes } from 'react';

export type OwnProps = {
  sign: string;
} & HTMLAttributes<HTMLDivElement>;
