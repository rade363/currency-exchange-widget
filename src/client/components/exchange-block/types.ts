import { Account } from '../../hooks/useExchange/types';

export type OwnProps = {
  accountFrom: Account;
  accountTo: Account;
  handleAccountFromInputChange: (value: string) => void;
  handleAccountToInputChange: (value: string) => void;
  swapAccounts: () => void;
  chooseFromAccount: () => void;
  chooseToAccount: () => void;
  transferMoney: (isButtonActive: boolean, isErrorPresent: boolean) => void;
};
