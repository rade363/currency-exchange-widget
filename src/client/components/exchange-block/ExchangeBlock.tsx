import React from 'react';
import AccountBlock from '../account-block';
import SwapButton from '../swap-button';
import Button from '../button';
import setButtonText from '../../utils/set-button-text';
import { OwnProps } from './types';
import './exchange-block.scss';

function ExchangeBlock(props: OwnProps) {
  const {
    accountFrom,
    accountTo,
    handleAccountFromInputChange,
    handleAccountToInputChange,
    swapAccounts,
    chooseFromAccount,
    chooseToAccount,
    transferMoney,
  } = props;
  const isErrorPresent = accountFrom.error !== null || accountTo.error !== null;
  const isButtonActive = Number(accountFrom.change) !== 0 && Number(accountTo.change) !== 0;

  return (
    <div className="exchange-block">
      <h1 className="exchange-block__title">
        Exchange {accountFrom.currency.name} to {accountTo.currency.name}
      </h1>
      <AccountBlock
        type="From"
        currentAccount={accountFrom}
        otherAccount={accountTo}
        onInputChange={handleAccountFromInputChange}
        handleSelectorClick={chooseFromAccount}
      />
      <SwapButton
        onClick={swapAccounts}
      />
      <AccountBlock
        type="To"
        currentAccount={accountTo}
        otherAccount={accountFrom}
        onInputChange={handleAccountToInputChange}
        handleSelectorClick={chooseToAccount}
      />
      <Button
        active={isButtonActive}
        styleType={isErrorPresent ? 'danger' : 'primary'}
        disabled={isErrorPresent}
        onClick={() => transferMoney(isButtonActive, isErrorPresent)}
      >
        {setButtonText(isErrorPresent, isButtonActive, accountFrom.currency.name, accountTo.currency.name)}
      </Button>
    </div>
  );
}

export default ExchangeBlock;
