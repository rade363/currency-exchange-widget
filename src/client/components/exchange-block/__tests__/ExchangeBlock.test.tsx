import React from 'react';
import { create } from 'react-test-renderer';
import ExchangeBlock from '../ExchangeBlock';
import { Account } from '../../../hooks/useExchange/types';

describe('Exchange block', () => {
  const accountFrom: Account = {
    balance: 1137.42,
    change: '0.00',
    result: null,
    error: null,
    currency: {
      name: 'USD',
      sign: '$',
      rate: 0.753425,
    },
  };

  const accountTo: Account = {
    balance: 1137.42,
    change: '0.00',
    result: null,
    error: null,
    currency: {
      name: 'GBP',
      sign: 'Â£',
      rate: 1.327273,
    },
  };

  test('Should match snapshot', () => {
    const mockCallBack = jest.fn();
    const block = create(
      <ExchangeBlock
        accountFrom={accountFrom}
        accountTo={accountTo}
        handleAccountFromInputChange={mockCallBack}
        handleAccountToInputChange={mockCallBack}
        swapAccounts={mockCallBack}
        chooseFromAccount={mockCallBack}
        chooseToAccount={mockCallBack}
        transferMoney={mockCallBack}
      />,
    );
    const tree = block.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
