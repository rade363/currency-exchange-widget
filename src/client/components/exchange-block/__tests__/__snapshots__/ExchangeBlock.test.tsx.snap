// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange block Should match snapshot 1`] = `
<div
  className="exchange-block"
>
  <h1
    className="exchange-block__title"
  >
    Exchange 
    USD
     to 
    GBP
  </h1>
  <div
    className="account-block margin_tb_s-5"
  >
    <div
      className="account-block__row"
    >
      <div
        className="account-block__type"
      >
        From
      </div>
      <div
        className="account-block__balance"
      >
        Balance: 
        1137.42
      </div>
    </div>
    <div
      className="account-block__row"
    >
      <button
        className="currency-label account-block__selector"
        onClick={[MockFunction]}
        type="button"
      >
        <div
          className="currency-sign currency-label__sign"
        >
          $
        </div>
        <div
          className="currency-label__name"
        >
          USD
        </div>
        <img
          alt="arrow-down"
          className="currency-label__arrow"
          src="test-file-stub"
        />
      </button>
      <input
        className="account-block__input"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="0.00"
        type="text"
        value=""
      />
    </div>
    <div
      className="account-block__row"
    >
      <div
        className="account-block__rate"
      >
        $
        1 = 
        £
        0.75
      </div>
      <div
        className="account-block__result "
      >
        
      </div>
    </div>
  </div>
  <button
    className="swap-button "
    onClick={[MockFunction]}
    type="button"
  >
    <span
      className="swap-button__arrows"
    >
      ↑↓
    </span>
  </button>
  <div
    className="account-block margin_tb_s-5"
  >
    <div
      className="account-block__row"
    >
      <div
        className="account-block__type"
      >
        To
      </div>
      <div
        className="account-block__balance"
      >
        Balance: 
        1137.42
      </div>
    </div>
    <div
      className="account-block__row"
    >
      <button
        className="currency-label account-block__selector"
        onClick={[MockFunction]}
        type="button"
      >
        <div
          className="currency-sign currency-label__sign"
        >
          £
        </div>
        <div
          className="currency-label__name"
        >
          GBP
        </div>
        <img
          alt="arrow-down"
          className="currency-label__arrow"
          src="test-file-stub"
        />
      </button>
      <input
        className="account-block__input"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="0.00"
        type="text"
        value=""
      />
    </div>
    <div
      className="account-block__row"
    >
      <div
        className="account-block__rate"
      >
        £
        1 = 
        $
        1.33
      </div>
      <div
        className="account-block__result "
      >
        
      </div>
    </div>
  </div>
  <button
    className="button button_primary  "
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Specify value
  </button>
</div>
`;
